### ISSUE TITLE
reactor spawns processes, calls processEnded, without actually being running

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5287]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5287|
|Type|defect|
|Created|2011-09-30 03:51:21Z|

This python program:
```
#!py
import os
from twisted.internet import reactor, protocol
class OneRun(protocol.ProcessProtocol):
  def processEnded(self, reason):
      print "Process ended."
reactor.spawnProcess(OneRun(), executable="echo", args=["echo", "hello"],
                     env=os.environ, childFDs={0:0, 1:1, 2:2})
import time
time.sleep(1)
```
... which you will note does _not_ actually run the reactor, causes "hello" and "Process ended." to be printed to stdout.  This indicates that the process is started, and ended, and the notification for it ending is delivered to the protocol.

I think the process has to be started right away, so that `spawnProcess` can have a complete return value, but  the `processEnded` hook should not be invoked unless the reactor is already running.

<details><summary>Searchable metadata</summary>

```
trac-id__5287 5287
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__ 
time__1317354681000000 1317354681000000
changetime__1409838768473385 1409838768473385
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Duplicate of [#5247](https://github.com/twisted/twisted/issues/5247)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

I tried this example on my Ubuntu 12.04  but could not reproduce it... process ended is called only after reactor is starts.


```
import os
from twisted.internet import reactor, protocol
class OneRun(protocol.ProcessProtocol):
  def processEnded(self, reason):
      print "Process ended."
reactor.spawnProcess(OneRun(), executable="echo", args=["echo", "hello"],
                     env=os.environ, childFDs={0:0, 1:1, 2:2})
import time
time.sleep(1)
print 'After sleep start reactor.'
reactor.callLater(1, reactor.stop)
reactor.run()
print 'Reactor stopped.'
```


output is

```
$ ./build-ubuntu1204-x64/bin/python bla.py 
hello
After sleep start reactor.
Process ended.
Reactor stopped.
```

-------

I am tried a similar test on Windows and 'process ended' is only called after reactor starts.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Note that in your test, you did see `Hello` before the sleep completed.

The behavior has changed slightly because we no longer install the SIGCHLD handler until the reactor is running - so even though we run the process and it exits before the reactor starts, we no longer *notice* that it exited until the reactor starts.

It is still probably a bad behavior that process is run before the reactor starts... but I guess that's not what glyph was actually complaining about.

I *think* the `processEnded` part of this was fixed in r28447 (though that may have been committed before this ticket was filed).  r28447 moved the SIGCHLD handler installation out of `ReactorBase.__init__` and into something triggered by `reactor.run`, anyway.
